### ISSUE TITLE
Process substitution does not work with bat

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/bat#80]
```
> cat <(echo hello)
hello

> bat <(echo hello)
───────┬───────────────────────────────────────────────────
       │ File: /proc/self/fd/11
───────┼───────────────────────────────────────────────────
───────┴───────────────────────────────────────────────────
```

The */proc/self/fd/11* filename is expected, but somehow we fail to read the file contents.

**Comment by eth-p** [comment]
On Mac OS X (and possibly other platforms), using a pipe will output nothing at all.

```shell
echo "Hello world" | bat
```

**Comment by D3Hunter** [comment]
use `strace` on `bat` to see whether it has read input from `stdio`

**Comment by sharkdp** [comment]
> On Mac OS X (and possibly other platforms), using a pipe will output nothing at all.
> 
> echo "Hello world" | bat

Reading from stdin is not supported in the latest release (only in master), see #2.

This bug is not related to reading from stdin. This is about reading from named pipes.

**Comment by BrainMaestro** [comment]
I did some debugging, and the issue is related to the `BufReader` It's failing to read anything for some reason, but you can read it fine from the regular `File` handle

**Comment by sharkdp** [comment]
Hm, I can not reproduce the bug with this small example:

``` rust
use std::env;
use std::fs::File;
use std::io::{BufReader, Read};

fn main() {
    let filename = env::args().nth(1).unwrap();
    println!("Reading from {}", filename);

    let file = File::open(filename).unwrap();
    let mut reader = BufReader::new(file);

    let mut contents = String::new();
    reader.read_to_string(&mut contents).unwrap();
    println!("Contents: {}", contents);
}
```

**Comment by BrainMaestro** [comment]
You're right. Something else is going on. If you try it with a longer file like `src/main.rs` in `bat`, it only prints part of the file. A huge chunk of it is lost. With process substitution, you can only read the file once, so maybe the file is being read before we can interact with it. I've checked most of the other steps, and my only other hunch is clap?

**Comment by sharkdp** [comment]
Wow, interesting. Thanks for investigating.

**Comment by sharkdp** [comment]
I think I know where this comes from. `syntect` is probably reading something from the file when trying to infer the language.

**Comment by sharkdp** [comment]
https://github.com/trishume/syntect/blob/6c6dc8a9a93f62af1210e8ed0983ffe27bc17b97/src/parsing/syntax_set.rs#L212-L214

**Comment by BrainMaestro** [comment]
Nice! I completely skipped that. How would this be handled though

**Comment by sharkdp** [comment]
We can check for FIFOs via https://doc.rust-lang.org/stable/std/os/unix/fs/trait.FileTypeExt.html#tymethod.is_fifo

**Comment by BrainMaestro** [comment]
Very cool.

**Comment by sharkdp** [comment]
Thank you for help. Your observation ("maybe the file is being read before we can interact with it") was spot-on!
